pkg/provider/provider.go

```diff
@@ -146,47 +145,52 @@ func (k *kubeProvider) DiffConfig(ctx context.Context, req *pulumirpc.DiffReques
 		}, nil
 	}
 
-	oldConfig, err := parseKubeconfigString(strings.Trim(olds["kubeconfig"].String(), "{}"))
+	var diffs, replaces []string
+
+	oldConfig, err := parseKubeconfigPropertyValue(olds["kubeconfig"])
 	if err != nil {
 		return nil, err
 	}
-	newConfig, err := parseKubeconfigString(strings.Trim(news["kubeconfig"].String(), "{}"))
+	newConfig, err := parseKubeconfigPropertyValue(news["kubeconfig"])
 	if err != nil {
 		return nil, err
 	}
 
-	var diffs []string
+	// Check for differences in provider overrides.
+	if !reflect.DeepEqual(oldConfig, newConfig) {
+		diffs = append(diffs, "kubeconfig")
+	}
+	if olds["context"] != news["context"] {
+		diffs = append(diffs, "context")
+	}
+	if olds["cluster"] != news["cluster"] {
+		diffs = append(diffs, "cluster")
+	}
 
 	// In general, it's not possible to tell from a kubeconfig if the k8s cluster it points to has
 	// changed. k8s clusters do not have a well defined identity, so the best we can do is check
-	// if the server endpoint has changed. This is not a foolproof method; a trivial counterexample
-	// is changing the loadbalancer or DNS entry pointing to the same cluster.
+	// if the settings for the active cluster have changed. This is not a foolproof method; a trivial
+	// counterexample is changing the load balancer or DNS entry pointing to the same cluster.
 	//
 	// Given this limitation, we try to strike a reasonable balance by planning a replacement iff
-	// the `clusters` or `current-context` configs in the kubeconfig change. This could still plan
-	// an erroneous replacement, but should work for the majority of cases.
+	// the active cluster in the kubeconfig changes. This could still plan an erroneous replacement,
+	// but should work for the majority of cases.
 	//
 	// The alternative of ignoring changes to the kubeconfig is untenable; if the k8s cluster has
 	// changed, any dependent resources must be recreated, and ignoring changes prevents that from
 	// happening.
-	if !reflect.DeepEqual(oldConfig.Clusters, newConfig.Clusters) ||
-		!reflect.DeepEqual(oldConfig.CurrentContext, newConfig.CurrentContext) {
-		diffs = append(diffs, "kubeconfig")
-	}
-
-	if !reflect.DeepEqual(olds["cluster"], news["cluster"]) {
-		diffs = append(diffs, "cluster")
+	oldActiveCluster := getActiveClusterFromConfig(oldConfig, olds)
+	activeCluster := getActiveClusterFromConfig(newConfig, news)
+	if !reflect.DeepEqual(oldActiveCluster, activeCluster) {
+		replaces = diffs
 	}
+	glog.V(7).Infof("%s: diffs %v / replaces %v\n", label, diffs, replaces)
```

Comment:
FYI: `glog.Infof` will add a trailing newline if one is not in the string, so you don't need to explicitly pass `\n`